timemachine/cpp/src/context.cu

```diff
@@ -147,54 +149,44 @@ std::array<std::vector<double>, 3> Context::multiple_steps_U(
         throw std::runtime_error("store_x_interval <= 0");
     }
 
+    if (n_steps % store_x_interval != 0) {
+        std::cout << "warning:: n_steps modulo store_x_interval does not equal zero" << std::endl;
+    }
+
     int n_windows = lambda_windows.size();
-    int u_traj_size = ((n_steps + store_u_interval - 1) / store_u_interval) * n_windows;
-    int x_traj_size = (n_steps + store_x_interval - 1) / store_x_interval;
+    int u_traj_size = (n_steps / store_u_interval) * n_windows;
+    int x_traj_size = (n_steps / store_x_interval);
     int box_traj_size = x_traj_size * 3 * 3;
 
     std::vector<double> h_x_traj(x_traj_size * N_ * 3);
 
+    // try catch block is to deal with leaks in d_u_buffer
     try {
         gpuErrchk(cudaMalloc(&d_box_traj, box_traj_size * sizeof(*d_box_traj)));
         gpuErrchk(cudaMalloc(&d_u_traj, u_traj_size * sizeof(*d_u_traj)));
         gpuErrchk(cudaMemset(d_u_traj, 0, u_traj_size * sizeof(*d_u_traj)));
 
-        for (int step = 0; step < n_steps; step++) {
+        cudaStream_t stream = static_cast<cudaStream_t>(0);
+        for (int step = 1; step <= n_steps; step++) {
+
+            this->_step(lambda, nullptr);
 
             if (step % store_x_interval == 0) {
                 gpuErrchk(cudaMemcpy(
-                    &h_x_traj[0] + (step / store_x_interval) * N_ * 3,
+                    &h_x_traj[0] + ((step / store_x_interval) - 1) * N_ * 3,
                     d_x_t_,
                     N_ * 3 * sizeof(double),
                     cudaMemcpyDeviceToHost));
                 gpuErrchk(cudaMemcpy(
-                    &d_box_traj[0] + (step / store_x_interval) * 3 * 3,
+                    &d_box_traj[0] + ((step / store_x_interval) - 1) * 3 * 3,
                     d_box_t_,
                     3 * 3 * sizeof(*d_box_traj),
                     cudaMemcpyDeviceToDevice));
             }
 
-            cudaStream_t stream = static_cast<cudaStream_t>(0);
-
-            gpuErrchk(cudaMemsetAsync(d_du_dx_t_, 0, N_ * 3 * sizeof(*d_du_dx_t_), stream));
-
-            // first pass generate the forces
-            for (int i = 0; i < bps_.size(); i++) {
-                bps_[i]->execute_device(
-                    N_,
-                    d_x_t_,
-                    d_box_t_,
-                    lambda,
-                    d_du_dx_t_, // we only need the forces
-                    nullptr,
-                    nullptr,
-                    nullptr,
-                    stream);
-            }
-
             // we need to compute aggregate energies on this step
```

Comment:
nit: is this comment still relevant since the first step was removed?

Reply:
I think so, the step that it does it changes but it is still aggregating all of the energies. Not the most informative comment though, so can remove if it seems confusing

Reply:
Maybe just reword to avoid "on this step"? It's a bit more confusing now because it seems like "this step" could refer to `step`

Reply:
Resolved in [dfb5198](https://github.com/proteneer/timemachine/pull/724/commits/dfb5198720efe6e9aed1a8c200f32d1926825f5a)